### ISSUE TITLE
codecov security incident

### ISSUE THREAD

**Issue by asmeurer** [sympy/sympy#21328]
codecov recently had a security incident https://about.codecov.io/security-update/.

I don't know if this could have affected us. I think the only thing that could have been leaked is the doctr deploy key, but as far as I can tell, that is on a different build in Travis than codecov. It may be prudent to recreate it either way (this is pretty easy, just install `doctr` and run `doctr configure` and follow the instructions).

We might consider removing codecov either way. Does anyone find it useful? I personally do not, and for SymPy, I've found it to be more annoying than anything given the various issues we've had with it https://github.com/sympy/sympy/issues?q=is%3Aopen+codecov+label%3ACoverage. In principle, showing coverage on pull requests is useful, but the particular implementation of this idea in codecov isn't a very good one.

**Comment by smichr** [comment]
> In principle, showing coverage on pull requests is useful

This, in addition to the tests that are added, give me a lot of confidence in viewing a PR favorably.

**Comment by oscarbenjamin** [comment]
Coverage is definitely useful. We could just set up Actions jobs to run the tests with coverage enabled and store the output from coverage in an artifact. We could also use something like this to get diff coverage:
https://github.com/Bachmann1234/diff_cover

That gives e.g. (for #21120):
```
$ pytest --cov --cov-report=xml sympy/polys/matrices/
$ diff-cover coverage.xml 
-------------
Diff Coverage
Diff: origin/master...HEAD, staged and unstaged changes
-------------
sympy/polys/matrices/domainmatrix.py (76.9%): Missing lines 779-780,782,792,796,798
sympy/polys/matrices/domainscalar.py (83.6%): Missing lines 54,60,66-69,76,82,86,90
sympy/polys/matrices/tests/test_domainmatrix.py (92.3%): Missing lines 550
sympy/polys/matrices/tests/test_domainscalar.py (100%)
-------------
Total:   147 lines
Missing: 17 lines
Coverage: 88%
-------------
```
It could also be possible to have the bot paste that on the PR or something which would be nice. Most important though is just for a reviewer to have some way of checking the coverage easily. This was a significant problem with codecov: the diff coverage didn't update after new commits were pushed.

**Comment by asmeurer** [comment]
@smichr @oscarbenjamin do you find codecov specifically useful? If you do, we can keep it. 

Having coverage generate an html report and diff report and pushing that to some gh-pages would probably end up being more useful than codecov, based on my experience. I don't know if that's hard to do with GitHub actions (specifically the pushing from PR builds). Even just having an artifact that you can manually download and open would be helpful, it sounds like.

**Comment by oscarbenjamin** [comment]
No, codecov wasn't very good. Unless it has improved I would try a different provider rather than reuse codecov.

Storing an artifact with diff coverage like I showed above is easy to do.

We could do the same for the benchmarks.

**Comment by oscarbenjamin** [comment]
We could try coveralls:
https://github.com/marketplace/actions/coveralls-python

**Comment by asmeurer** [comment]
I've heard it's not much better, although I don't see any issues with trying it. We should make sure it's on an isolated job from anything has secret keys to avoid issues like the above one.

**Comment by oscarbenjamin** [comment]
I just checked reviews and they don't seem to suggest that it's much better either. I do wonder whether part of the issue with codecov for sympy is just the size of the codebase. It was always very slow and updated very intermittently. Maybe it just struggles when faced with a million lines of code.
